addons/hw_drivers/controllers/driver.py

```diff
@@ -116,22 +120,30 @@ def connect_box(self, token):
 # Drivers
 #----------------------------------------------------------
 
-drivers = []
+drivers = deque()
 bt_devices = {}
 socket_devices = {}
 iot_devices = {}
 
-class DriverMetaClass(type):
+class DriverMetaClass(ABCMeta):
     def __new__(cls, clsname, bases, attrs):
         newclass = super(DriverMetaClass, cls).__new__(cls, clsname, bases, attrs)
-        drivers.append(newclass)
+        #  We only add concrete classes to `drivers`.
+        if inspect.isabstract(newclass):
+            return newclass
+        # Some drivers must be tried only when all the others have been ruled out. These are kept at the bottom of the list.
+        if newclass.is_tested_last:
+            drivers.append(newclass)
```

Comment:
what if two drivers are `is_tested_last`? wouldn't a sequence be more appropriate?

Reply:
It would, but we don't intend to keep this very long, and there won't be new scale drivers until then. In the near future we'll let the user choose a driver for each port in a wizard.